CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/RawSyntaxNodesFile.swift

```diff
@@ -103,15 +106,69 @@ func rawSyntaxNodesFile(nodesStartingWith: [Character]) -> SourceFileSyntax {
             }
 
-            for (swiftName, kind) in choices {
+            for (caseName, upperCaseName, kind) in choices {
               if let choiceNode = SYNTAX_NODE_MAP[kind], choiceNode.kind.isBase {
                 DeclSyntax(
                   """
-                  public init(\(swiftName): some \(choiceNode.kind.rawProtocolType)) {
-                    self = .\(swiftName)(\(choiceNode.kind.rawType)(\(swiftName)))
+                  public init(\(caseName.declarationSiteNameOrCallSiteVarName): some \(choiceNode.kind.rawProtocolType)) {
+                    self = .\(caseName.callSiteMemberName)(\(choiceNode.kind.rawType)(\(caseName.declarationSiteNameOrCallSiteVarName)))
                   }
                   """
                 )
+
+                DeclSyntax(
+                  """
+                  public func is\(upperCaseName)(type \(caseName)Type: (some \(choiceNode.kind.rawProtocolType)).Type = \(choiceNode.kind.rawType).self) -> Bool {
+                    self.as(\(caseName)Type) != nil
+                  }
+                  """
+                )
+
+                DeclSyntax(
+                  """
+                  public func as\(upperCaseName)<S: \(choiceNode.kind.rawProtocolType)>(type: S.Type = \(choiceNode.kind.rawType).self) -> S? {
+                    if case .\(caseName.callSiteMemberName)(let \(caseName.declarationSiteNameOrCallSiteVarName)) = self {
+                      return S(\(caseName.declarationSiteNameOrCallSiteVarName))
+                    } else {
+                      return nil
+                    }
+                  }
+                  """
+                )
+
+                DeclSyntax(
+                  """
+                  public func cast\(upperCaseName)<S: \(choiceNode.kind.rawProtocolType)>(type: S.Type = \(choiceNode.kind.rawType).self) -> S {
+                    self.as(S.self)!
+                  }
+                  """
+                )
+              } else {
+                try VariableDeclSyntax(
+                  "public var is\(upperCaseName): Bool"
+                ) {
+                  ExprSyntax("self.as\(upperCaseName) != nil")
+                }
+
+                try VariableDeclSyntax(
+                  "public var as\(upperCaseName): \(OptionalTypeSyntax(wrappedType: kind.rawType))"
+                ) {
+                  StmtSyntax(
+                    """
+                    if case .\(caseName.callSiteMemberName)(let \(caseName.declarationSiteNameOrCallSiteVarName)) = self {
+                      return \(caseName.declarationSiteNameOrCallSiteVarName)
+                    } else {
+                      return nil
+                    }
+                    """
+                  )
+                }
+
+                try VariableDeclSyntax(
+                  "public var cast\(upperCaseName): \(kind.rawType)"
+                ) {
+                  ExprSyntax("self.as\(upperCaseName)!")
+                }
```

Comment:
I think we only use one of these functions in a single place. I’m not sure if it’s worth generating them.